File: kotlin/solutions/01-beginner/ex033_property_custom_getter/Rectangle.kt
package fewolearning.exercises.beginner.ex033_property_custom_getter

/*
Exercise 033 - Property custom getter (reference solution).
*/
class Rectangle(val width: Int, val height: Int) {
    val area: Int
        get() = width * height

    val perimeter: Int
        get() = 2 * (width + height)
}

File: php/exercises/01-beginner/ex034_basic_class_constructor/Rectangle.php
<?php

declare(strict_types=1);

namespace FeWoLearning\Exercises\Beginner\Ex034BasicClassConstructor;

/*
Exercise 034 - Constructor property promotion (beginner).

Goal:   Implement a Rectangle using constructor property promotion and computed methods.
Drills: constructor property promotion, computed methods over promoted properties.
Passes: RectangleTest
*/
final class Rectangle
{
    public function __construct(
        private readonly float $width,
        private readonly float $height
    ) {
        if ($width < 0.0 || $height < 0.0) {
            throw new \InvalidArgumentException('Width and height must not be negative.');
        }
    }

    public function area(): float
    {
        throw new \RuntimeException('TODO');
    }

    public function perimeter(): float
    {
        throw new \RuntimeException('TODO');
    }
}
